src/Gui/Inventor/Draggers/Gizmo.cpp

```diff
@@ -0,0 +1,697 @@
+ *   License along with FreeCAD. If not, see                                *
+ *   <https://www.gnu.org/licenses/>.                                       *
+ *                                                                          *
+ ***************************************************************************/
+
+#include "PreCompiled.h"
+
+#include "Gizmo.h"
+
+#ifndef _PreComp_
+#include <Inventor/nodes/SoOrthographicCamera.h>
+#include <Inventor/nodes/SoPerspectiveCamera.h>
+#include <Inventor/nodes/SoPickStyle.h>
+#include <algorithm>
+#endif
+
+#include <App/Application.h>
+#include <Base/Converter.h>
+#include <Base/Parameter.h>
+#include <Base/Precision.h>
+#include <Document.h>
+#include <Gui/Inventor/So3DAnnotation.h>
+#include <Gui/Inventor/SoToggleSwitch.h>
+#include <Gui/QuantitySpinBox.h>
+#include <Gui/Utilities.h>
+#include <Gui/View3DInventorViewer.h>
+
+#include "SoLinearDragger.h"
+#include "SoLinearDraggerGeometry.h"
+#include "SoRotationDragger.h"
+#include "SoRotationDraggerGeometry.h"
+
+using namespace Gui;
+
+void Gizmo::setDraggerPlacement(const Base::Vector3d& pos, const Base::Vector3d& dir)
+{
+    setDraggerPlacement(
+        Base::convertTo<SbVec3f>(pos),
+        Base::convertTo<SbVec3f>(dir)
+    );
+}
+
+bool Gizmo::delayedUpdateEnabled()
+{
+    static Base::Reference<ParameterGrp> hGrp = App::GetApplication()
+        .GetUserParameter()
+        .GetGroup("BaseApp/Preferences/Mod/PartDesign");
+
+    return hGrp->GetBool("DelayedGizmoUpdate", false);
+}
+
+LinearGizmo::LinearGizmo(QuantitySpinBox* property)
+{
+    setProperty(property);
+}
+
+SoInteractionKit* LinearGizmo::initDragger()
+{
+    draggerContainer = new SoLinearDraggerContainer;
+    draggerContainer->color.setValue(1, 0, 0);
+    dragger = draggerContainer->getDragger();
+
+    dragger->addStartCallback(
+        [] (void* data, SoDragger*) {
+            assert(data);
+            static_cast<LinearGizmo*>(data)->draggingStarted();
+        },
+        this
+    );
+    dragger->addFinishCallback(
+        [] (void* data, SoDragger*) {
+            assert(data);
+            static_cast<LinearGizmo*>(data)->draggingFinished();
+        },
+        this
+    );
+    dragger->addMotionCallback(
+        [] (void* data, SoDragger*) {
+            assert(data);
+            static_cast<LinearGizmo*>(data)->draggingContinued();
+        },
+        this
+    );
+
+    dragger->labelVisible = false;
+
+    setDragLength(property->value().getValue());
+
+    dragger->instantiateBaseGeometry();
+
+    // change the dragger dimensions
+    auto arrow = SO_GET_PART(dragger, "arrow", SoArrowGeometry);
+    arrow->cylinderHeight = 3.5;
+    arrow->cylinderRadius = 0.2;
+
+    return draggerContainer;
+}
+
+void LinearGizmo::uninitDragger()
+{
+    dragger = nullptr;
+    draggerContainer = nullptr;
+}
+
+GizmoPlacement LinearGizmo::getDraggerPlacement()
+{
+    assert(draggerContainer);
```

Comment:
It probably should be an if with early return and perhaps warning. With the assert application will crash anyway.

Reply:
Yeah that is the intent here. This assert should only hit if someone is using my Gizmo API wrongly. It is impossible to hit for an user unless the API is used erroneously. So, we want to tell the developer right away about it by crashing. I can added a proper message though to make it more helpful.